hazelcast-sql/src/main/java/com/hazelcast/sql/impl/expression/FieldAccessExpression.java

```diff
@@ -76,33 +76,31 @@ public T eval(Row row, ExpressionEvalContext context) {
     @Override
     public T eval(final Row row, final ExpressionEvalContext context, boolean useLazyDeserialization) {
         // Use lazy deserialization for nested queries. Only the last access should be eager.
-        final Object res = ref.eval(row, context, true);
-        if (res == null) {
+        final Object result = ref.eval(row, context, true);
+        if (result == null) {
             return null;
         }
 
-        if (isPrimitive(res.getClass())) {
+        if (isPrimitive(result.getClass())) {
             throw QueryException.error("Field Access expression can not be applied to primitive types");
         }
 
-        if (getterCache.get() == null) {
-            getterCache.compareAndSet(null, new EvictableGetterCache(
+        if (getterCache == null) {
```

Comment:
It is not thread-safe. Volatile modifier grants write atomicity, and prevents compiler to  reorder reads and writes to this reference. What you've got here - the `EvictableGetterCache`  may be created multiple times, because several threads may enter the condition body simultaneously. I'd recommend you to rollback this change.

Reply:
This version has the same synchronization characteristics as the previous one. `AtomicReference.compareAndSet` does not take a supplier, so `EvictableGetterCache` can still be created multiple times.

Reply:
It's wrong. `compareAndSet` **atomically** compares and set new value to the atomic ref.

Reply:
What you pass to `compareAndSet` is an `EvictableGetterCache` instance, not `Supplier<EvictableGetterCache>`. If multiple threads execute `getterCache.compareAndSet(...)` at the same time, multiple instances of `EvictableGetterCache` will be created, one of which will win.

Reply:
What you pass to `compareAndSet` is an `EvictableGetterCache` instance, not `Supplier<EvictableGetterCache>`. If multiple threads execute `getterCache.compareAndSet(...)` at the same time, multiple instances of `EvictableGetterCache` will be created, one of which will win. It is equivalent to the following.

```java
if (getterCache.get() == null) {
    GetterCache cache = new EvictableGetterCache(...);
    getterCache.compareAndSet(null, cache);
}
```